Initial implementation of the metadata log

varqox opened the pull request.

tropuq commented on include/seastar/fs/path.hh:
```diff
@@ -0,0 +1,37 @@
+
+#include <seastar/core/sstring.hh>
+
+namespace seastar {
```
```suggestion
namespace seastar::fs {
```

tropuq commented on include/seastar/fs/path.hh:
```diff
@@ -0,0 +1,37 @@
+/*
```
I think it would be better to move path.hh to src.

tropuq commented on src/fs/metadata_log.cc:
```diff
@@ -0,0 +1,323 @@
+future<> metadata_log::flush_curr_cluster_and_change_it_to_new_one() {
+    auto next_cluster = _cluster_allocator.alloc();
+    if (not next_cluster) {
+        // Here metadata log dies, we cannot even flush current cluster because from there we won't be able to recover
```
Maybe add TODO: to that comment to find that place easier later.

tropuq commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,252 @@
+
+namespace seastar::fs {
+
+struct invalid_inode_exception : public std::exception {
```
It would be better to have a base class for those exceptions so we can catch filesystem-specific exceptions easier. Also we could add error code to those exceptions (see std::filesystem::filesystem_error).

varqox replied:
What about those error codes, use ones from <cerrno> or invent new ones (in fact, there are not enough specific codes in <cerrno> for our needs).

psarna replied:
Let's just start by creating a base class for fs exceptions, and leave specific error codes for a later discussion - we'll probably use a mix of standard error codes and some additional errors specific for our implementation.

tropuq commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,252 @@
+    void write_update(inode_info::file& file, inode_data_vec new_data_vec);
+
+    // Deletes data vectors that are subset of @p data_range and cuts overlapping data vectors to make them not overlap
+    void cut_out_data_range(inode_info::file& file, file_range range);
```
I think that method could be private.

varqox replied:
It should be, but how we test it then?

varqox replied:
My bad, it indeed can be private.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+    metadata_log& _metadata_log;
+    cluster_range _available_clusters;
+    std::unordered_set<cluster_id_t> _taken_clusters;
+    std::optional<inode_t> _next_cluster;
```
```suggestion
    std::optional<cluster_id_t> _next_cluster;
```

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+
+    future<> bootstrap(cluster_id_t first_metadata_cluster_id, fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id) {
+        _next_cluster = first_metadata_cluster_id;
+        return do_with((cluster_id_t)first_metadata_cluster_id, [this](cluster_id_t& last_cluster) {
```
Is there a reason for that (cluster_id_t)?

varqox replied:
Unfortunately, do_with() needs rvalue and that is how it is created here.

psarna replied:
It's fine, but let's use constructor style instead of a C-style cast, which are forbidden in seastar - `cluster_id_t(first_metadata_cluster_id)`.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+            });
+        }).then([this, fs_shards_pool_size, fs_shard_id] {
+            // Initialize _cluser_allocator
+            std::deque<cluster_id_t> free_clusters;
```
We need to add clusters with medium and big writes to _taken_clusters. I would add a TODO for that.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    static auto invalid_entry_exception() {
+        return make_exception_future<>(std::runtime_error("Invalid metadata log entry"));
```
Add some details about an error to the message of invalid_entry_exception.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+        }
+
+        boost::crc_32_type crc;
+        if (not _curr_cluster.process_crc_without_reading(crc, checkpoint.checkpointed_data_length)) {
```
Probably we would like to have some debug logs to find out easily that something went wrong during bootstrapping.

psarna replied:
Indeed. Example:
```cpp
#include <seastar/util/log.hh>
logger mlogger("fs_metadata");
...
mlogger.warn("Something went wrong: {}", something);
```

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    future<> bootstrap_next_metadata_cluster() {
+        ondisk_next_metadata_cluster entry;
```
We allow for multiple NEXT_METADATA_CLUSTER entries. I think that it is invalid so I would report an error in that case.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+            return invalid_entry_exception();
+        }
+
+        _metadata_log._inodes[entry.inode].metadata = ondisk_metadata_to_metadata(entry.metadata);
```
Maybe in order to be consistent use something like _metadata_log.update_metadata(entry.inode, entry.metadata).

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+        }
+        temporary_buffer<uint8_t>& data = *data_opt;
+
+        if (not std::holds_alternative<inode_info::file>(_metadata_log._inodes[entry.inode].contents)) {
```
We can move that check before reading the data.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+            return invalid_entry_exception(); // Only unlinked inodes may be deleted
+        }
+
+        if (std::holds_alternative<inode_info::directory>(inode_info.contents) and not std::get<inode_info::directory>(inode_info.contents).entries.empty()) {
```
Maybe add functions like is_directory(inode), is_file(inode).

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+            return invalid_entry_exception();
+        }
+
+        // Only files may be linked as not to create cycles
```
We cannot create a directory first and then add it to another directory. Maybe instead of completely forbidding directories we could forbid adding directory that is already linked somewhere?

varqox replied:
So far only create_inode_as_dir_entry is allowed to be used to create directories. I don't think we want to allow / need linking directories.(add_dir_entry will be used for linking files, create_inode_as_dir_entry is used to create and link simultaneously.

tropuq replied:
So I would at least add a comment about that decision because for now, it is a little confusing I think.

psarna replied:
Hardlinks for directories are almost never implemented in file systems (never seen one), since they only create trouble, so I'm fine with leaving the code as is - the comment above that it's forbidden because it could create cycles is good enough.

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+        if (not std::holds_alternative<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents)) {
+            return invalid_entry_exception();
+        }
+        auto& dir = std::get<inode_info::directory>(_metadata_log._inodes[entry.dir_inode].contents);
```
Maybe add functions like get_directory(inode), get_file(inode).

tropuq commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,485 @@
+            return invalid_entry_exception();
+        }
+
+        if (_metadata_log._inodes.at(it->second).directories_containing_file == 0) {
```
It looks like it should be 'assert' instead of 'if'.

varqox replied:
Right. That assert is already in memory_only_delete_dir_entry(), so I will remove this if completely.

tropuq commented on src/fs/metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,249 @@
+
+#pragma once
+
+#include "bitwise.hh"
```
```suggestion
#include "fs/bitwise.hh"
```
Also, apply that suggestion to other relevant places.

tropuq commented on src/fs/metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,249 @@
+    using to_disk_buffer::reset;
+
+    /**
+     * @brief Clears buffer, leaving it in state as if it was just constructed
```
I think that it doesn't leave metadata_to_disk_buffer in any state that is reachable directly after constructing metadata_to_disk_buffer. In that case, that comment is misleading.

varqox replied:
I changed the comment to be more appropriate.

tropuq commented on src/fs/metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,249 @@
+     * @param metadata_end_pos position at which valid metadata ends: valid metadata range: [0, @p metadata_end_pos)
+     * @param align_before_buffering whether to align to beginning of the next unflushed data or set it to @p metadata_end_pos
+     */
+    void reset_from_bootstraped_cluster(disk_offset_t cluster_beg_offset, const uint8_t* cluster_contents, size_t metadata_end_pos, bool align_before_buffering) noexcept {
```
```suggestion
    void reset_from_bootstrapped_cluster(disk_offset_t cluster_beg_offset, const uint8_t* cluster_contents, size_t metadata_end_pos, bool align_before_buffering) noexcept {
```

tropuq commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,252 @@
+                        return make_ready_future<stop_iteration>(stop_iteration::yes); // Directory disappeared
+                    }
+                    if (not std::holds_alternative<inode_info::directory>(it->second.contents)) {
+                        return make_ready_future<stop_iteration>(stop_iteration::yes); // Directory became a file (should not happen, but safe-check)
```
I think it is not probable, but possible so I would remove that part "should not happen..." because it could be misleading and suggest that it is impossible unless there is a bug in code.

varqox replied:
I removed "(should not happen, but safe-check)" part.

tropuq commented on src/fs/to_disk_buffer.hh:
```diff
@@ -0,0 +1,139 @@
+    size_t _zero_padded_end; // Optimization to skip padding before write if it is already done
+
+public:
+    // Represents buffer that will be written to a block_device at offset @p disk_alligned_write_offset. Total number of bytes appended cannot exceed @p aligned_max_size.
```
```suggestion
    // Represents buffer that will be written to a block_device at offset @p disk_aligned_write_offset. Total number of bytes appended cannot exceed @p aligned_max_size.
```

tropuq commented on src/fs/to_disk_buffer.hh:
```diff
@@ -0,0 +1,139 @@
+public:
+    virtual void append_bytes(const void* data, size_t len) noexcept {
+        assert(len <= bytes_left());
+        memcpy(_buff.get_write() + _unflushed_data.end, data, len);
```
Use std::memcpy and apply that suggestion to other relevant places.

psarna commented on src/fs/inode_info.hh:
```diff
@@ -0,0 +1,152 @@
+
+    struct directory {
+        // TODO: directory entry cannot contain '/' character --> add checks for that
+        std::map<sstring, inode_t> entries; // entry name => inode
```
nitpick: we're in the process of purging sstring in favour of std::string (https://github.com/scylladb/seastar/issues/634), so let's use std::string in all new projects, including our FS.

psarna commented on src/fs/inode_info.hh:
```diff
@@ -0,0 +1,152 @@
+        std::map<file_offset_t, inode_data_vec> data; // file offset => data vector that begins there (data vectors do not overlap)
+
+        file_offset_t size() const noexcept {
+            return (data.empty() ? 0 : (--data.end())->second.data_range.end);
```
a safer and more human-readable version of `--data.end()` is `data.rbegin()`

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -0,0 +1,326 @@
+        0,
+        metadata,
+        [&]() -> decltype(inode_info::contents) {
+            if (is_directory)
```
please always wrap `if` body in curly braces

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -0,0 +1,326 @@
+        }
+
+        // TODO: I don't like that we make a copy here -- it is totally redundant and inhibits adding noexcept
+        sstring component = path.substr(component_range.beg, component_range.size());
```
1. Please switch to std::string, as mentioned above
2. How about std::string_view then? Using it will postpone the copy until it's really needed (e.g. the substring needs to be put somewhere)

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -0,0 +1,326 @@
+    return components_stack.back();
+}
+
+future<inode_t> metadata_log::futurized_path_lookup(const sstring& path) const {
```
This is the function we're more likely to use than raw `path_lookup`, right? I suggest naming this one `path_lookup`, and the other one `do_path_lookup`, it's a de-facto standard in seastar.

psarna commented on src/fs/metadata_log.cc:
```diff
@@ -0,0 +1,326 @@
+    return create_file_operation::perform(*this, std::move(path), std::move(perms), true).discard_result();
+}
+
+// TODO: think about how to make filesystem recoverable from ENOSPACE situation: flush() (or something else) throws ENOSPACE, then it should be possible to compact some data (e.g. by truncating a file) via top-level interface and retrying the flush() without a ENOSPACE error. In particular if we delete all files after ENOSPACE it should be successful. It becomes especially hard if we write metada ...
```
Warm congratulations on a 557 character line! 🎉 Please break it a little, since I got bored scrolling and I still don't know the last 25% of this comment :)

psarna commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,275 @@
+
+    void memory_only_add_dir_entry(inode_info::directory& dir, inode_t entry_inode, sstring entry_name);
+
+    void memory_only_delete_dir_entry(inode_info::directory& dir, sstring entry_name);
```
nitpick: it's fine for the declarations above to be separated by a single newline only, not double.

psarna commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,488 @@
+    , _curr_cluster_data(decltype(_curr_cluster_data)::aligned(metadata_log._alignment, metadata_log._cluster_size))
+    {}
+
+    future<> bootstrap(cluster_id_t first_metadata_cluster_id, fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id) {
```
Note: the code below looks fine, but consider simply spawning a seastar thread with `seastar::async`. Bootstrapping is a rare operation and `seastar::async` makes the code so much more readable, plus cancels the need of `do_with`s and other liveness-related tricks.

varqox replied:
Does seastar::async have any performance impact beyond small memory footprint? If so, I think that the code is in pretty good shape.

psarna replied:
`seastar::async` needs to allocate a rather large stack - currently 128KiB, but it's going to be customizable soon. On the other hand, the stack is allocated once, so it might sometimes be beneficial, compared to a complicated chain of futures, each requesting a small allocation.

Once we have coroutines, that would be the third viable alternative - which will have one small allocation for all needed variables.

psarna commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,488 @@
+            // Reset _inode_allocator
+            std::optional<inode_t> max_inode_no;
+            if (not _metadata_log._inodes.empty()) {
+                max_inode_no = (--_metadata_log._inodes.end())->first;
```
again - `rbegin` is slightly more readable than `--end()`

psarna commented on src/fs/metadata_log_bootstrap.hh:
```diff
@@ -0,0 +1,488 @@
+
+    future<> bootstrap_rename_dir_entry() {
+        // TODO: implement it
+        assert(false && "Not implemented");
```
About all these asserts - these are reachable for clients, right? Let's return an exception future instead, so that the process doesn't crash when calling a function - it will make testing easier.

varqox replied:
For now we don't intend to use / test this functions before implementing them, so I thought that assert is fine.

psarna replied:
ok

psarna commented on src/fs/metadata_log_operations/create_file.hh:
```diff
@@ -0,0 +1,127 @@
+        });
+    }
+
+    future<inode_t> perform_3() {
```
That's the third worst function name I've ever seen, and that's only because I scrolled over `perform_1` and `perform_2` first :smile: But, back to being serious, please add descriptive names - they may be long, but they should express what the function does.

varqox replied:
The problem with naming here is that the flow is somewhat inverted. perform_1 returns future resolving when whole operation is complete. perform_2 returns the same, but assumes that path parsing is done and global lock is acquired. perform_3 returns the same, but assumes what perform_2 assumed and also that an unique lock is acquired on the directory entry to create. I don't have a good idea how to name it, maybe as if the flow was not inverted?

psarna replied:
It's fine for the order to be inverted and mixed, it's quite common for seastar apps. Let's just name functions based on what they do - if `perform_1` is the main entry point, so `create_file` sounds a little obvious, but it's still a good name. `perform_2` assumes we looked up a dir inode, so something along `create_file_in_directory` will explain its purpose. Finally, `create_file_in_directory` is just a thin wrapper to get the directory and its lock, so `perform_3` can be named `do_create_file_in_directory`, since that's what actually happens in there.

`perform_2` and `perform_3` are only used internally, so the users wouldn't care if their order is inverted or not - but it's important to name them in a self-explanatory way, so that reviewers and developers know what was the intent.

psarna commented on src/fs/metadata_log_operations/create_file.hh:
```diff
@@ -0,0 +1,127 @@
+        unix_metadata unx_mtdt = {
+            _perms,
+            getuid(), // TODO: this or something else?
+            getgid(), // TODO: this or something else?
```
0 is also a fine placeholder for now, but getuid and getgid are natural candidates, so it's fine, although local uid/gid do not provide much context for a distributed system. Eventually, we'll want a user to be able to pass his credentials when bootstrapping the file system - that will allow us to authorize users on startup (e.g. via LDAP or whatnot). You can add a TODO comment about it around here.

psarna reviewed (commented):
Thanks, looks very good after a cursory review. I'll need more time to review it properly, but I left initial comments anyway.

Is it correct that unit tests for `cut_out_data_range` are not part of this pull request? If so, please add some - it's a fragile part of code, ranges are always error-prone.

Also, please go over the series and add some text to commit messages - we usually only merge patches without descriptions (title only) if they are trivial (e.g. fixing a typo), while this series is rather complex. Instead of marking some titles with `[see details]` (which can by the way be removed), assume that all commit messages will have details inside them - take a look at seastar commit history to see examples of what info can be put in the description.

The "apply numerous suggestions..." commit will need to be rebased out, but I assume it just waits here until the review is finished and will be rebased-out afterwards, which is fine.

Good job!

tropuq commented on src/fs/inode_info.hh:
```diff
@@ -0,0 +1,152 @@
+            }
+
+            while (it != data.end() and are_intersecting(range, it->second.data_range)) {
+                auto data_vec = std::move(it->second);
```
You can use std::map::extract

varqox commented on src/fs/metadata_log_operations/create_file.hh:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+
+        _entry_name = last_component(path);
```
@psarna Is it OK, to use std::filesystem::path(path).filename().string() instead of last_component(path)?. It would make last_component() redundant, but I don't know the performance implications of using std::filesystem::path...

varqox replied:
I dug a little through the libstdc++ source code and it seems that std::filesystem::path splits path on every '/' and stores path both as a string and vector of its components that have type std::filesystem::path. So it seems a lot of unnecessary copying.

psarna replied:
It's still fine for now - we leave microoptimizations for later and focus on core file system implementation. If it ever comes up in perf as a bottleneck, it will be easy to fix.

varqox commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,278 @@
+
+namespace seastar::fs {
+
+struct fs_exception : public std::exception {
```
Maybe we would like to export all exceptions to seastar/include/fs since they will propagate up up to the user?

wmitros commented on src/fs/metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,250 @@
+
+private:
+    bool fits_for_append(size_t bytes_no) const noexcept {
+        return (bytes_left() >= bytes_no + sizeof(ondisk_next_metadata_cluster)); // We need to reserve space for the next metadata cluster entry
```
```suggestion
        return (bytes_left() >= bytes_no + sizeof(ondisk_type) + sizeof(ondisk_next_metadata_cluster)); // We need to reserve space for the next metadata cluster entry
```

varqox commented:
Now this PR depends on #80.

tropuq commented on src/fs/path.hh:
```diff
@@ -0,0 +1,42 @@
+namespace seastar::fs {
+
+// Extracts the last component in @p path. WARNING: The last component is empty iff @p path is empty or ends with '/'
+std::string extract_last_component(std::string& path) {
```
```suggestion
inline std::string extract_last_component(std::string& path) {
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,316 @@
+
+    template<typename T>
+    bool is_append_type(size_t idx) {
+        return is_type<action::append>(idx) && std::holds_alternative<T>(get_by_type<action::append>(idx).entry);
```
```suggestion
        return is_type<action::append>(idx) and std::holds_alternative<T>(get_by_type<action::append>(idx).entry);
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,316 @@
+        _unflushed_data = {real_write.end, real_write.end};
+        if (bytes_left() > 0) {
+            start_new_unflushed_data();
+        }
```
```suggestion
        start_new_unflushed_data();
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,316 @@
+    }
+
+private:
+    void move_bytes_count(size_t len) {
```
```suggestion
    void mock_append_bytes(size_t len) {
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,316 @@
+
+    void prepare_unflushed_data_for_flush() noexcept override {}
+
+    append_result check_and_move_bytes_count(size_t data_len) {
```
```suggestion
    append_result mock_append(size_t data_len) {
```

varqox commented on src/fs/metadata_disk_entries.hh:
```diff
@@ -186,4 +186,33 @@ struct ondisk_rename_dir_entry_header {
 } __attribute__((packed));
 
+template<typename T>
+constexpr size_t get_ondisk_entry_size(const T& entry) noexcept {
```
maybe
```suggestion
constexpr size_t ondisk_entry_size(const T& entry) noexcept {
```
?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer.hh:
```diff
@@ -0,0 +1,316 @@
+    }
+
+private:
+    temporary_buffer<uint8_t> copy_data(const void* data, size_t length) {
```
maybe something like
```suggestion
    static temporary_buffer<uint8_t> make_tmp_buff(const void* data, size_t length) {
```
?

varqox replied:
or
```suggestion
    static temporary_buffer<uint8_t> as_tmp_buff(const void* data, size_t length) {
```
?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/test_runner.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <stdint.h>
```
```suggestion
#include <cstdint>
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+    constexpr size_t next_metadata_cluster_size = sizeof(ondisk_type) + sizeof(ondisk_next_metadata_cluster);
+
+    size_t remaining_space = default_buff_size - checkpoint_size - next_metadata_cluster_size;
+    while (true) {
```
Maybe you would want to use this:
```suggestion
    for (;;) {
```
for infinite loops?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+    BOOST_REQUIRE_EQUAL(buf.append(ondisk_truncate {64, 28, 62}), APPENDED);
+    BOOST_REQUIRE_EQUAL(buf.append(ondisk_mtime_update {4, 26}), APPENDED);
+    buf.flush_to_disk(dev).get();
+
```
There is no check for the contents of added entries. Was it intended?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+    auto buf2 = buf.virtual_constructor(default_buff_size, default_alignment);
+    buf2->init(default_alignment);
+
+    BOOST_REQUIRE_EQUAL(created_buffers.size(), 1);
```
I think it is somewhat misleading that created_buffers holds only buffers created by call to virtual_constructor(). I only realized it thanks to this test. Maybe a better name would solve this problem, I can only come up with e.g. cloned_buffers, virtually_constructed_buffers.

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+    buf2->append(ondisk_delete_inode {1});
+    BOOST_REQUIRE_EQUAL(created_buffers[0]->actions.size(), 1);
+
+    auto buf3 = buf2->virtual_constructor(default_buff_size, default_alignment);
```
Maybe it is a good idea to move it and add buf4 here above buf2->append, so that we check that only buf2 is appended and add the same checks that ale below, but for buf4?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+
+SEASTAR_THREAD_TEST_CASE(update_metadata_test) {
+    mock_metadata_to_disk_buffer buf(default_buff_size, default_alignment);
+    buf.init(0);
```
This repeats everywhere, maybe make a function like `default_mock_buf` for that?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+            get_ondisk_entry_size(update_metadata_op));
+
+    BOOST_CHECK(buf.is_append_type<ondisk_update_metadata>(0));
+    auto &elem = buf.get_by_append_type<ondisk_update_metadata>(0);
```
This also appears everywhere. Maybe a function that does all that and takes the current operation as an argument and a lambda which will be called with `elem` as an argument would simplify it a little. Checks about elem would be placed in that lambda. What do you think?

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+        42,
+        24,
+        static_cast<uint16_t>(old_name_str.size()),
+        static_cast<uint16_t>(new_name_str.size())};
```
```suggestion
        static_cast<uint16_t>(new_name_str.size())
    };
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+    std::string create_inode_str = "120345";
+    ondisk_create_inode_as_dir_entry_header create_inode_op {
+            {42, 1, {5, 2, 6, 8, 4}}, 7, static_cast<uint16_t>(create_inode_str.size())
+        };
```
Maybe?
```suggestion
        {42, 1, {5, 2, 6, 8, 4}}, 7, static_cast<uint16_t>(create_inode_str.size())
    };
```

varqox commented on tests/unit/fs_mock_metadata_to_disk_buffer_test.cc:
```diff
@@ -0,0 +1,398 @@
+        24,
+        static_cast<uint16_t>(old_name_str.size()),
+        static_cast<uint16_t>(new_name_str.size())};
+    BOOST_REQUIRE_EQUAL(buf.append(rename_op, old_name_str.c_str(), new_name_str.c_str()), APPENDED);
```
I think data() is better suited here (bonus: they are equivalent in semantics).
```suggestion
    BOOST_REQUIRE_EQUAL(buf.append(rename_op, old_name_str.data(), new_name_str.data()), APPENDED);
```
If you accept it, change it also in the above tests.

varqox reviewed (commented):
@tropuq Here you have a review.

varqox commented on src/fs/metadata_log.hh:
```diff
@@ -211,6 +212,10 @@ public:
             fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id);
 
 private:
+    bool inode_exists(inode_t inode) const {
```
```suggestion
    bool inode_exists(inode_t inode) const noexcept {
```

varqox commented on src/fs/inode_info.hh:
```diff
@@ -136,6 +136,10 @@ struct inode_info {
 
     std::variant<directory, file> contents;
 
+    bool is_unlinked() const {
```
```suggestion
    bool is_unlinked() const noexcept {
```

varqox commented on tests/unit/fs_metadata_log_test.cc:
```diff
@@ -76,14 +76,21 @@ SEASTAR_THREAD_TEST_CASE(mock_metadata_to_disk_buffer_test) {
+    log.create_directory("/test/", file_permissions::default_dir_permissions).get();
+    log.create_file("/test/test", file_permissions::default_dir_permissions).get();
 
     auto &created_buffers = mock_metadata_to_disk_buffer::created_buffers;
```
I think it is a prefered way of making a reference. If you agree, please change in other places also.
```suggestion
    auto& created_buffers = mock_metadata_to_disk_buffer::created_buffers;
```

varqox commented on src/fs/metadata_log.cc:
```diff
@@ -348,6 +348,39 @@ future<inode_t> metadata_log::open_file(std::string path) {
+                // Inode removal operation is started in background
+                return _locks.with_lock(metadata_log::locks::unique {inode}, [this, inode, file_info] {
+                    if (not inode_exists(inode) or not file_info->is_unlinked()) {
+                        return now();
```
Add comment e.g.
```suggestion
                        return now(); // Scheduled delete became invalid
```

psarna commented:
I merged first 3 commits (after squashing 2 fixups as well), and continue the review, successfully merging stuff or commenting on found issues.

psarna commented on src/fs/inode_info.hh:
```diff
@@ -0,0 +1,223 @@
+
+    struct directory {
+        // TODO: directory entry cannot contain '/' character --> add checks for that
+        std::map<std::string, inode_t, std::less<>> entries; // entry name => inode
```
Any particular reason for providing `std::less` explicitly? It's the default.

varqox replied:
Default is `std::less<Key>`. The advantage of `std::less<void>` aka `std::less<>` is that it declares member `is_transparent` which enables `template<class K> std::map::find(const K& x) const` that allows search for element via "key" of different type than `std::map::key_type` without converting it to the `std::map::key_type` thus omitting unnecessary copies and allocations e.g. if `K` is `std::string_view` and `key_type` is `std::string` (this is the case is used in `metadata_log::path_lookup()`.

varqox replied:
Default is `std::less<Key>`. The advantage of `std::less<void>` aka `std::less<>` is that it declares member `is_transparent` which enables `template<class K> std::map::find(const K& x) const` that allows search for element via "key" of different type than `std::map::key_type` without converting it to the `std::map::key_type` thus omitting unnecessary copies and allocations e.g. if `K` is `std::string_view` and `key_type` is `std::string` (this is the case is used in `metadata_log::path_lookup()`.

psarna replied:
Ok, heterogenous lookup, got it

psarna commented on src/fs/inode_info.hh:
```diff
@@ -0,0 +1,223 @@
+    constexpr file& get_file() & noexcept { return std::get<file>(contents); }
+    constexpr const file& get_file()  const & noexcept { return std::get<file>(contents); }
+    constexpr file&& get_file() && noexcept { return std::move(std::get<file>(contents)); }
+    constexpr const file&& get_file() const && noexcept { return std::move(std::get<file>(contents)); }
```
Why would you ever need a const rvalue reference? I think these overloads just bloat the interface and are never actually useful. While at it, it's more human-readable when you omit the ampersand in `const & noexcept {` - this is the default and people are not used to seeing these.

To sum up, I'd rather `get_file` (and, respectively, `get_directory`) was declared as:
```cpp
    file& get_file() & noexcept { return std::get<file>(contents); }
    const file& get_file() const noexcept { return std::get<file>(contents); }
    file&& get_file() && noexcept { return std::move(std::get<file>(contents)); }
```
If you prefer (or need) to explicitly declare any of these as constexpr then I'm fine with that, but they can be safely omitted as well.

varqox replied:
This gives compilation error:
```cpp

struct bla {
	void foo() & {}
	void foo() const {}
	void foo() && {}
};

int main() {}
```
```gcc
/tmp/a.cc:3:7: error: ‘void bla::foo() const’ cannot be overloaded with ‘void bla::foo() &’
    3 |  void foo() const {}
      |       ^~~
/tmp/a.cc:2:7: note: previous declaration ‘void bla::foo() &’
    2 |  void foo() & {}
      |       ^~~
```

psarna replied:
Ok, you're right. `&` stays, so please just get rid of the unnecessary `const file&&` overload.

psarna commented on src/fs/metadata_log_operations/unlink_or_remove_file.hh:
```diff
@@ -0,0 +1,176 @@
+
+            switch (_metadata_log.append_ondisk_entry(ondisk_entry, _entry_name.data())) {
+            case metadata_log::append_result::TOO_BIG:
+                assert(false and "ondisk entry cannot be too big");
```
These kinds of errors also deserve throwing, not asserting. If we're unable to write an entry, there are graceful ways of shutting down the server, crashing it is too harsh.

varqox replied:
I think it is a bug that we cannot append entry to a fresh (empty) cluster -- it is too small to append this entry => it is too small for the file-system to function properly. So i treated it as a bug -- added assert for that. Maybe such checking (with exceptions) should be done in metadata_to_disk_buffer's constructor, but for me it is a bug here, where assert is situated.

psarna replied:
Ok, I'm fine with assertions as long as the input is properly validated earlier.

varqox replied:
I turns out this entry doesn't have fixed size, so indeed it should be an exception because we cannot check for it beforehand. Thx

psarna commented on src/fs/metadata_log.hh:
```diff
@@ -0,0 +1,395 @@
+    void memory_only_delete_dir_entry(inode_info::directory& dir, std::string entry_name);
+
+    template<class Func>
+    void schedule_background_task(Func&& task) {
```
This queue of background tasks is unbounded, which is worrying - effectively we can have an unbounded parallelism for background operations, and it's easy to overload/OOM the server that way.

This one would be better if it was:
1. Coordinated by an external "background worker" service
2. Throttled

This service can simply be a seastar thread which has a big loop which is guarded with a parallelism-bounding semaphore and simply takes stuff from a queue and runs it. To provide a pseudo-seastar example:

```cpp
class background_worker {
    static constexpr int max_parallelism = 1000;
    semaphore _sem{max_parallelism};
    some_kind_of_queue _background_tasks_to_run;
    abort_source _as;
    future<> _run_worker_thread;
public:
    background_worker () : _run_worker_thread(start_worker()) {}

    future<> stop() {
        _as.request_abort();
       _sem.broken();
        return std::move(_started);
    }

    future<> enqueue_task(some_task_type&& task) {
        return _background_tasks_to_run.push(std::move(task));
    }

protected:
    future<> start_worker() {
        return async [] {
            while (!_as.abort_requested()) {
                wait_for_the_semaphore_to_reduce_parallelism().get();
                wait_until_the_queue_is_not_empty().get();
                auto task_to_run = _background_tasks_to_run.pop_front();
                task_to_run().get();
            }
        }
    }
}
```

and then, you just start the background worker as part of the file system runtime for each shard, and then `schedule_background_task` can be implemented as:
```cpp
future<> schedule_background_task(Func&& task) {
    return _background_worker.enqueue_task(std::move(task));
}
```
Note that `schedule_background_task` returns a future, which reduces the unbounded parallelism of background tasks.

psarna commented on src/fs/metadata_log_operations/create_file.hh:
```diff
@@ -0,0 +1,138 @@
+    inode_t _dir_inode;
+    inode_info::directory* _dir_info;
+
+    create_file_operation(metadata_log& metadata_log) : _metadata_log(metadata_log) {}
```
This constructor needs to initialize `_is_directory` to something, otherwise debug mode warns that sometimes a non-initialized value is used here. You can also put
```cpp
bool _is_directory = true;
```
in the above declaration and be done with all esoteric cases.

varqox committed:
fs: metadata_log: add stat() operation

Provides inteface to query file attributes that include permissions,
btime, mtime and ctime.

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

wmitros committed:
tests: fs: add truncate operation test

Checks whether the data that will be written to disk after truncate is correct,
the reads from a truncated file are accurate and the files metadata is set
to the new size.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

varqox committed:
fs: metadata_log: add base implementation

SeastarFS is a log-structured filesystem. Every shard will have 3
private logs:
- metadata log
- medium data log
- big data log (this is not actually a log, but in the big picture it
  looks like it was)

Disk space is divided into clusters (typically around several MiB) that
have all equal size that is multiple of alignment (typically 4096
bytes). Each shard has its private pool of clusters (assignment is
stored in bootstrap record). Each log consumes clusters one by one -- it
writes the current one and if cluster becomes full, then log switches to
a new one that is obtained from a pool of free clusters managed by
cluster_allocator. Metadata log and medium data log write data in the
same manner: they fill up the cluster gradually from left to right. Big
data log takes a cluster and completely fills it with data at once -- it
is only used during big writes.

This commit adds the skeleton of the metadata log:
- data structures for holding metadata in memory with all operations on
  this data structure i.e. manipulating files and their contents
- locking logic (detailed description can be found in metadata_log.hh)
- buffers for writting logs to disk (one for metadata and one for medium
  data)
- basic higher level interface e.g. path lookup, iterating over
  directory
- boostraping metadata log == reading metadata log from disk and
  reconstructing shard's filesystem structure from just before shutdown

File content is stored as a set of data vectors that may have one of
three kinds: in memory data, on disk data, hole. Small writes are
writted directly to the metadata log and because all metadata is stored
in the memory these writes are also in memory, therefore in-memory kind.
Medium and large data are not stored in memory, so they are represented
using on-disk kind. Enlarging file via truncate may produce holes, hence
hole kind.

Directory entries are stored as metadata log entries -- directory inodes
have no content.

To disk buffers buffer data that will be written to disk. There are two
kinds: (normal) to disk buffer and metadata to disk buffer. The latter
is implemented using the former, but provides higher level interface for
appending metadata log entries rather than raw bytes.

Normal to disk buffer appends data sequentially, but if a flush occurs
the offset where next data will be appended is aligned up to alignment
to ensure that writes to the same cluster are non-overlaping.

Metadata to disk buffer appends data using normal to disk buffer but
does some formatting along the way. The structure of the metadata log on
disk is as follows:
| checkpoint_1 | entry_1, entry_2, ..., entry_n | checkpoint_2 | ... |
               | <---- checkpointed data -----> |
etc. Every batch of metadata_log entries is preceded by a checkpoint
entry. Appending metadata log appends the current batch of entries.
Flushing or lack of space ends current batch of entries and then
checkpoint entry is updated (because it holds CRC code of all
checkpointed data) and then write of the whole batch is requested and a
new checkpoint (if there is space for that) is started. Last checkpoint
in a cluster contains a special entry pointing to the next cluster that
is utilized by the metadata log.

Bootstraping is, in fact, just replying of all actions from metadata log
that were saved on disk. It works as follows:
- reads metadata log clusters one by one
- for each cluster, until the last checkpoint contains pointer to the
  next cluster, processes the checkpoint and entries it checkpoints
- processing works as follows:
  - checkpoint entry is read and if it is invalid it means that the
    metadata log ends here (last checkpoint was partially written or the
    metadata log really ended here or there was some data corruption...)
    and we stop
  - if it is correct, it contains the length of the checkpointed data
    (metadata log entries), so then we process all of them (error there
    indicates that there was data corruption but CRC is still somehow
    correct, so we abort all bootstraping with an error)

Locking is to ensure that concurrent modifications of the metadata do
not corrupt it. E.g. Creating a file is a complex operation: you have
to create inode and add a directory entry that will represent this inode
with a path and write corresponding metadata log entries to the disk.
Simultaneous attempts of creating the same file could corrupt the file
system. Not to mention concurrent create and unlink on the same path...
Thus careful and robust locking mechanism is used. For details see
metadata_log.hh.

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

varqox committed:
fs: metadata_log: add operation for creating and opening unlinked file

Creating unlinked file may be useful as temporary file or to expose the
file via path only after the file is filled with contents.

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

varqox committed:
fs: metadata_log: add creating files and directories

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

varqox committed:
fs: metadata_log: add private operation for deleting inode

Some operations need to schedule deleting inode in the background. One
of these is closing unlinked file if nobody else holds it open.

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

varqox committed:
fs: metadata_log: add link operation

Allows the same file to be visible via different paths or to give a path
to an unlinked file.

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

varqox committed:
fs: metadata_log: add unlinking files and removing directories

Signed-off-by: Krzysztof Małysa <varqox@gmail.com>

tropuq committed:
fs: metadata_log: add opening file

Marks that the file is opened by increasing the opened file counter.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: metadata_log: add closing file

Decreases opened file counter. If the file is unlinked and the
counter is zero then the file is automatically removed.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: metadata_log: add write operation

Each write can be divided into multiple smaller writes that can fall
into one of the following categories:
- small write: writes below SMALL_WRITE_THRESHOLD bytes, those writes
  are stored fully in memory
- medium write: writes above SMALL_WRITE_THRESHOLD and below
  cluster_size bytes, those writes are stored on disk, they are appended
  to the on-disk data log where data from different writes can be stored
  in one cluster
- big write: writes that fully fit into one cluster, stored on disk
For example, one write can be divided into multiple big writes, some
small writes and some medium writes. Current implementation won't make
any unnecessary data copying. Data given by caller is either directly
used to write to disk or is copied as a small write.

Added cluster writer which is used to perform medium writes. Cluster
writer keeps a current position in the data log and appends new data
by writing it directly into disk.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

wmitros committed:
fs: metadata_log: add truncate operation

Truncate can be used on a file to change its size. When the new
size is lower than current, the data at higher offsets will be lost,
and when it's larger, the file will be filled with null bytes.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

tropuq committed:
fs: metadata_log: add read operation

Reads file data from disk and memory based on information stored in
inode's data vectors. Not optimized version - reads from disk are always
read into temporary buffers before copying to the buffer given by the
caller.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

wmitros committed:
tests: fs: add to_disk_buffer test

The test checks whether the data written by a to_disk_buffer to disk
is the same as the data appended to the buffer and the remaining buffer
space is correctly calculated on small examples.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

tropuq committed:
tests: fs: added metadata_to_disk_buffer and cluster_writer mockers

Added mockers:
- mockers store information about every operation
- store list of virtually created mockers

Added tests for metadata_to_disk_buffer mocker. Tests check that
mocker behaves similarly to metadata_to_disk_buffer.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
tests: fs: add write test

- random tests
- tests for corner cases
  * basic single small writes
  * basic single medium writes
  * basic single large writes
  * new cluster allocation for medium writes
  * medium write split into two smaller writes due to lack of space in
    data-log cluster
  * split single write into more smaller writes because of unaligned
    buffer
  * split big write (bigger than cluster size) into multiple writes

Signed-off-by: Michał Niciejewski <quport@gmail.com>

wmitros committed:
tests: fs: add metadata_to_disk_buffer unit tests

For every ondisk entry check if:
- it's correctly appended to the buffer when it would fit
- the buffer returns TOO_BIG when it wouldn't fit
- it's written to disk after successful append and flush.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

tropuq committed:
fs: read: add optimization for aligned reads

Optimization for aligned reads. When on-disk data and given buffer are
properly aligned than read disk data is not stored in a temporary
buffer but is directly read into the buffer given by the caller.

Added device_reader to perform unaligned reads with caching.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
tests: fs: add tests for aligned reads and writes

Random test checking aligned writes and reads optimizations.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

rokinsky committed:
tests: fs: add basic test for metadata log bootstrapping

Checks if there is access to the newly created directories after bootstrapping.

Signed-off-by: Aleksander Sorokin <ankezy@gmail.com>
